lib/puppet/provider/cisco_itd_device_group_node/cisco.rb

```diff
@@ -0,0 +1,229 @@
+  ITDDG_NODE_BOOL_PROPS = [
+    :hot_standby,
+    :probe_control,
+  ]
+  ITDDG_NODE_ALL_PROPS = ITDDG_NODE_NON_BOOL_PROPS + ITDDG_NODE_BOOL_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@nu',
+                                            ITDDG_NODE_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@nu',
+                                            ITDDG_NODE_BOOL_PROPS)
+
+  def initialize(value={})
+    super(value)
+    itddg = @property_hash[:itddg]
+    node = @property_hash[:node]
+    @nu = Cisco::ItdDeviceGroupNode.itd_nodes[itddg][node] unless
+      itddg.nil? || node.nil?
+    @property_flush = {}
+  end
+
+  def self.properties_get(itd_device_group_name, node, nu_obj)
+    debug "Checking instance, #{itd_device_group_name} #{node}"
+    current_state = {
+      name:      "#{itd_device_group_name} #{node}",
+      itddg:     itd_device_group_name,
+      node:      node,
+      node_type: nu_obj.node_type,
+      ensure:    :present,
+    }
+    # Call node_utils getter for each property
+    ITDDG_NODE_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = nu_obj.send(prop)
+    end
+
+    ITDDG_NODE_BOOL_PROPS.each do |prop|
+      val = nu_obj.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    itd_nodes = []
+    Cisco::ItdDeviceGroupNode.itd_nodes.each do |itddg, all_nu|
+      all_nu.each do |nu, nu_obj|
+        itd_nodes << properties_get(itddg, nu, nu_obj)
+      end
+    end
+    itd_nodes
+  end
+
+  def self.prefetch(resources)
+    itd_nodes = instances
+    resources.keys.each do |id|
+      provider = itd_nodes.find do |node|
+        node.itddg.to_s == resources[id][:itddg].to_s &&
+        node.node.to_s == resources[id][:node].to_s &&
+        node.node_type.to_s == resources[id][:node_type].to_s
+      end
+      resources[id].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    (@property_hash[:ensure] == :present)
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def instance_name
+    node
+  end
+
+  def properties_set(new_itd_node=false)
+    ITDDG_NODE_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_itd_node
+      unless @property_flush[prop].nil?
+        @nu.send("#{prop}=", @property_flush[prop]) if
+          @nu.respond_to?("#{prop}=")
+      end
+    end
+    # custom setters which require one-shot multi-param setters
+    probe_set
+    hot_standby_weight_set
+  end
+
+  # We need special handling for boolean properties in our custom
+  # setters below. This helper method returns true if the property
+  # flush contains a TrueClass or FalseClass value.
+  def flush_boolean?(prop)
```

Comment:
Given that we are now using this in multiple providers, please move this into 

https://github.com/cisco/cisco-network-puppet-module/blob/efc8885cc8971b56514558d42b509de30d57bd47/lib/puppet_x/cisco/cmnutils.rb#L22